File: blog/20140601/gini_impurity/gini.hs
import Data.List

size = fromIntegral . length

-- (a) 1 - (AA + BB + CC)
giniA xs = (1 - ) . sum . map calc . group . sort $ xs
	where
		listSize = size xs
		calc x = (size x / listSize) ** 2

-- (b) AB + AC + BA + BC + CA + CB
giniB xs = sum . calcProb . map prob . group . sort $ xs
	where
		listSize = size xs
		prob ys = (head ys, size ys / listSize)
		calcProb zs = [ snd x * snd y | x <- zs, y <- zs, fst x /= fst y]

main = do
	let list = ["A", "B", "B", "C", "B", "A"]

	print $ giniA list
	print $ giniB list

File: blog/20140608/gini_impurity2/gini.fs

let size xs = xs |> Seq.length |> float

// (a) 1 - (AA + BB + CC)
let giniA xs = xs |> Seq.countBy id |> Seq.sumBy (fun (k, v) -> (float v / size xs) ** 2.0) |> (-) 1.0

let combinationCount cs = [
    for x in cs do
        for y in cs do
            if fst x <> fst y then
                yield (snd x, snd y)
]

// (b) AB + AC + BA + BC + CA + CB
let giniB xs = xs |> Seq.countBy id |> combinationCount |> Seq.sumBy (fun (x, y) -> (float x / size xs) * (float y / size xs))

let list = ["A"; "B"; "B"; "C"; "B"; "A";]

printfn "%A" (giniA list)
printfn "%A" (giniB list)
